File: section02_rn_bridge_ios/example_bridge_ios/ios/exampleModule/RCTExampleModule.m
// Importa o header do sistema de bridge do React Native.
// Esse arquivo contém a definição do protocolo RCTBridgeModule,
// usado para registrar módulos nativos que podem ser acessados pelo JavaScript.
#import "React/RCTBridgeModule.h"

// Declaração de um módulo nativo exposto ao React Native.
// A macro RCT_EXTERN_MODULE informa ao React Native que existe
// uma classe nativa chamada "RCTExampleModule", herdando de NSObject.
//
// Importante: a implementação real desse módulo está em Swift.
// Aqui estamos apenas expondo a assinatura do módulo para o runtime Objective-C,
// que é quem o React Native usa para registrar módulos.
@interface RCT_EXTERN_MODULE(RCTExampleModule, NSObject)

// Expõe o método printMessage ao JavaScript.
// A macro RCT_EXTERN_METHOD indica que esse método existe na implementação Swift
// e deve ser acessível via NativeModules no JS.
//
// No JavaScript, ele poderá ser chamado assim:
// NativeModules.RCTExampleModule.printMessage();
RCT_EXTERN_METHOD(printMessage: (NSString *) title: (NSNumber *) value)
RCT_EXTERN_METHOD(
  returnMessage:
  (NSString *)title
  resolver:
  (RCTPromiseResolveBlock)resolve
  rejecter:
  (RCTPromiseRejectBlock)reject
)
RCT_EXTERN_METHOD(eventMessage: (CGFloat *) value)

@end

File: section02_rn_bridge_ios/example_bridge_ios/ios/exampleModule/RCTExampleModule.swift
import Foundation
import React

@objc(RCTExampleModule)
class RCTExampleModule: NSObject {
  
  /// Exemplo de módulo nativo exposto ao React Native.
  /// A anotação @objc permite que o módulo e seus métodos
  /// sejam acessados via JavaScript.
  ///
  /// O método `printMessage` pode ser chamado pelo lado
  /// JavaScript e executará um código nativo em Swift.
  
  @objc
  func printMessage(_ title: String, _ value: NSNumber) {
    print("Olá, \(title), Você tem \(value) anos.")
  }
  
  @objc
  func returnMessage(
    _ title: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    if title == "Lucas" {
      resolve("Message processed: \(title)")
    } else {
      let error = NSError(domain: "", code: 200, userInfo: [NSLocalizedDescriptionKey: "Titutlo de erro"])
      reject("Error", "Mensagem de erro", error)
    }
  }
}
